tests/benchmark/ExecutorBench.cpp

```diff
@@ -0,0 +1,395 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "glow/Backends/DeviceManager.h"
+#include "glow/Runtime/Executor/Executor.h"
+
+#include "ExecutorBenchmarkWrapper.h"
+#include "ExecutorTestBuilder.h"
+#include "TestDeviceManager.h"
+
+#include "benchmark/benchmark.h"
+
+using namespace glow;
+using namespace glow::runtime;
+
+/// Fixture class for reusing Executor, ExecutorTestBuilder and DeviceManagerMap
+/// setup across benchmarks.
+class ThreadPoolExecutorBench : public benchmark::Fixture {
+public:
+  ThreadPoolExecutorBench()
+      : executor_(std::shared_ptr<Executor>(
+            createExecutor(deviceManagerMap_, ExecutorKind::ThreadPool))),
+        benchmarkBuilder_(executor_, deviceManagerMap_) {}
+  ~ThreadPoolExecutorBench() = default;
+
+protected:
+  /// The Executor being tested.
+  std::shared_ptr<Executor> executor_;
+  /// An ExecutorTestBuilder instance for creating benchmarks.
+  ExecutorTestBuilder benchmarkBuilder_;
+  /// DeviceManager map for initializing executor_.
+  DeviceManagerMapTy deviceManagerMap_;
+};
+
+/// Benchmark for executing an empty DAG.
+BENCHMARK_DEFINE_F(ThreadPoolExecutorBench, Empty)(benchmark::State &state) {
+  constexpr RunIdentifierTy testRunId = 10;
+
+  for (auto _ : state) {
+    // Pause timing while setting up to run the benchmark.
+    state.PauseTiming();
+
+    // Make a PlaceholderBindings with one Placeholder in it to make sure
+    // Executor::run() doesn't modify it when the root given to it is null. Make
+    // two identical copies; one to give to Executor::run(), and another to
+    // compare the returned PlaceholderBindings with.
+    PseudoRNG rng;
+    auto type = std::unique_ptr<Type>(new Type(ElemKind::FloatTy, {1, 2, 2}));
+    auto placeholder = llvm::make_unique<Placeholder>("a", type.get(),
+                                                      /*trainable=*/false);
+
+    auto testContext = llvm::make_unique<ExecutionContext>();
+    auto refContext = llvm::make_unique<ExecutionContext>();
+
+    auto *tensor =
+        testContext->getPlaceholderBindings()->allocate(placeholder.get());
+    tensor->init(Tensor::InitKind::Xavier, 1.0, rng);
+    refContext->getPlaceholderBindings()->insert(placeholder.get(),
+                                                 tensor->clone());
+
+    // Variables for storing runId actually returned by
+    // Executor::run() via its callback.
+    RunIdentifierTy executorRunId;
+    std::unique_ptr<ExecutionContext> executorOutputContext;
+
+    // Call Executor::run().
+    llvm::Error runErr = llvm::Error::success();
+    std::promise<void> promise;
+    std::future<void> future = promise.get_future();
+
+    // Start timing.
+    state.ResumeTiming();
+    executor_->run(nullptr, std::move(testContext), testRunId,
```

Comment:
Can you use the benchmark wrapper run here to avoid duplicating code?

Reply:
Hmm, I'll try.